broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/DbSnapshotStore.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.broker.clustering.atomix.storage.snapshot;
+
+import io.atomix.protocols.raft.storage.snapshot.PendingSnapshot;
+import io.atomix.protocols.raft.storage.snapshot.Snapshot;
+import io.atomix.protocols.raft.storage.snapshot.SnapshotListener;
+import io.atomix.protocols.raft.storage.snapshot.SnapshotStore;
+import io.atomix.utils.time.WallClockTimestamp;
+import io.zeebe.broker.logstreams.state.StatePositionSupplier;
+import io.zeebe.util.FileUtil;
+import io.zeebe.util.ZbLogger;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.CopyOnWriteArraySet;
+import org.slf4j.Logger;
+
+public class DbSnapshotStore implements SnapshotStore {
+  private static final Logger LOGGER = new ZbLogger(DbSnapshotStore.class);
+
+  // if thread-safe is a must, then switch to ConcurrentNavigableMap
+  // a map of all available snapshots indexed by index
+  private final ConcurrentNavigableMap<DbSnapshotId, DbSnapshot> snapshots;
+  // the root snapshotsDirectory where all snapshots should be stored
+  private final Path snapshotsDirectory;
+  // the root snapshotsDirectory when pending snapshots should be stored
+  private final Path pendingDirectory;
+  // keeps track of all snapshot modification listeners
+  private final Set<SnapshotListener> listeners;
+  // used to fetch the position for snapshots replicated by Atomix
+  private final StatePositionSupplier positionSupplier;
+  // a pair of mutable snapshot ID for index-only lookups
+  private final ReusableSnapshotId lowerBoundId;
+  private final ReusableSnapshotId upperBoundId;
+
+  public DbSnapshotStore(
+      final Path snapshotsDirectory,
+      final Path pendingDirectory,
+      final ConcurrentNavigableMap<DbSnapshotId, DbSnapshot> snapshots) {
+    this.snapshotsDirectory = snapshotsDirectory;
+    this.pendingDirectory = pendingDirectory;
+    this.snapshots = snapshots;
+
+    this.positionSupplier = new StatePositionSupplier(-1, LOGGER);
+    this.lowerBoundId = new ReusableSnapshotId(Long.MIN_VALUE);
+    this.upperBoundId = new ReusableSnapshotId(Long.MAX_VALUE);
+    this.listeners = new CopyOnWriteArraySet<>();
+  }
+
+  public DbSnapshot put(final DbSnapshot snapshot) {
+    // caveat: if the metadata is the same but the location is different, this will do nothing
```

Comment:
when can this happen?

Reply:
It shouldn't but 

![bugs](https://user-images.githubusercontent.com/43373/69816164-20ea2980-11f8-11ea-8a4f-51b01a2945ab.jpg)

Reply:
:laughing: